src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/CollectTraceActionTests.cs

```diff
@@ -0,0 +1,159 @@
+            {
+                rootOptions.AddFileSystemEgress(ExpectedEgressProvider, tempDirectory.FullName);
+
+                rootOptions.CreateCollectionRule(DefaultRuleName)
+                    .AddCollectTraceAction(ExpectedProviders, ExpectedEgressProvider, out CollectTraceOptions collectTraceOptions)
+                    .SetStartupTrigger();
+
+                collectTraceOptions.Duration = TimeSpan.FromSeconds(2);
+            }, async host =>
+            {
+                await PerformTrace(host);
+            });
+        }
+
+        private async Task PerformTrace(IHost host)
+        {
+            IOptionsMonitor<CollectionRuleOptions> ruleOptionsMonitor = host.Services.GetService<IOptionsMonitor<CollectionRuleOptions>>();
+            CollectTraceOptions options = (CollectTraceOptions)ruleOptionsMonitor.Get(DefaultRuleName).Actions[0].Settings;
+
+            ICollectionRuleActionFactoryProxy factory;
+            Assert.True(host.Services.GetService<ICollectionRuleActionOperations>().TryCreateFactory(KnownCollectionRuleActions.CollectTrace, out factory));
+
+            EndpointInfoSourceCallback callback = new(_outputHelper);
+            await using var source = _endpointUtilities.CreateServerSource(out string transportName, callback);
+            source.Start();
+
+            AppRunner runner = _endpointUtilities.CreateAppRunner(transportName, TargetFrameworkMoniker.Net60); // Arbitrarily chose Net60
+
+            Task<IEndpointInfo> newEndpointInfoTask = callback.WaitForNewEndpointInfoAsync(runner, CommonTestTimeouts.StartProcess);
+
+            await runner.ExecuteAsync(async () =>
+            {
+                IEndpointInfo endpointInfo = await newEndpointInfoTask;
+
+                ICollectionRuleAction action = factory.Create(endpointInfo, options);
+
+                using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(CommonTestTimeouts.TraceTimeout);
+
+                CollectionRuleActionResult result;
+                try
+                {
+                    await action.StartAsync(cancellationTokenSource.Token);
+
+                    result = await action.WaitForCompletionAsync(cancellationTokenSource.Token);
+                }
+                finally
+                {
+                    await DisposableHelper.DisposeAsync(action);
+                }
+
+                Assert.NotNull(result.OutputValues);
+                Assert.True(result.OutputValues.TryGetValue(CollectionRuleActionConstants.EgressPathOutputValueName, out string egressPath));
+                Assert.True(File.Exists(egressPath));
+
+                using FileStream traceStream = new(egressPath, FileMode.Open, FileAccess.Read);
+                Assert.NotNull(traceStream);
+
+                await ValidateTrace(traceStream, cancellationTokenSource.Token); // Do we want to be using a different token here? Does our timeout matter for this?
```

Comment:
```
                await ValidateTrace(traceStream, cancellationTokenSource.Token); // Do we want to be using a different token here? Does our timeout matter for this?
```

Yes, because capturing the trace would have taken a chunk of the time already. Rather than reusing the token source here, inside the ValidateTrace method, you can create a new CancellationTokenSource with a reasonable timeout e.g. 30 seconds. The validation shouldn't take more than a few seconds, if that, since the file stream should represent a completed trace.